src/primitives/transaction.h

```diff
@@ -425,11 +468,39 @@ struct CMutableTransaction
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(this->nVersion);
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwintered and nVersion.
+            uint32_t header;
+            READWRITE(header);
+            fOverwintered = header >> 31;
+            this->nVersion = header & 0x7FFFFFFF;
+        } else {
+            // When serializing v1 and v2, the 4 byte header is nVersion
+            uint32_t header = this->nVersion;
+            // When serializing Overwintered tx, the 4 byte header is the combination of fOverwintered and nVersion
+            if (fOverwintered) {
+                header |= 1 << 31;
+            }
+            READWRITE(header);
+        }
         nVersion = this->nVersion;
+        if (fOverwintered) {
+            READWRITE(nVersionGroupId);
+        }
+
+        bool isOverwinterV3 = fOverwintered &&
+                              nVersionGroupId == OVERWINTER_VERSION_GROUP_ID &&
+                              nVersion == 3;
+        if (fOverwintered && !isOverwinterV3) {
+            throw std::ios_base::failure("Unknown transaction format");
+        }
+
         READWRITE(vin);
         READWRITE(vout);
         READWRITE(nLockTime);
+        if (isOverwinterV3) {
+            READWRITE(nExpiryHeight);
+        }
```

Comment:
All the near-duplicate code here bothers me.

Reply:
It will be fixed at the same time the `const_cast` stuff is.

Reply:
@daira I can't add a comment to your comment above.  No, exceptions during deserialization are rare but there are classes that do this.  I found between 5-10 across the code base, example:
```
template <typename Stream, typename Operation>
    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
        unsigned char leadingByte = 0x00;
        READWRITE(leadingByte);

        if (leadingByte != 0x00) {
            throw std::ios_base::failure("lead byte of NotePlaintext is not recognized");
        }

        READWRITE(value);
        READWRITE(rho);
        READWRITE(r);
        READWRITE(memo);
    }
```

Reply:
I keep writing comments about this and then being unable to find them...

Exceptions are not rare in terms of parser coverage, because `ReadCompactSize` throws on non-canonical encodings, so every single serialization type that uses a compact size (string, vector, map, set, ...) has the ability to throw.

Reply:
In that case it's fine to throw here.